crates/cadenza-gcode/src/gcode_syntax.rs

```diff
@@ -0,0 +1,263 @@
+//! GCode lexer and parser that produces Cadenza-compatible AST.
+//!
+//! This module treats GCode as an alternative syntax for Cadenza. It lexes and parses
+//! GCode into a Rowan CST that can be directly evaluated by the Cadenza eval crate.
+//!
+//! # Architecture
+//!
+//! - **Lexer**: Tokenizes GCode (G1, X100, comments, etc.)
+//! - **Parser**: Builds Rowan GreenNode CST using cadenza-syntax token kinds
+//! - **AST**: GCode commands become Apply nodes, parameters become literals
+//!
+//! # Example
+//!
+//! ```
+//! use cadenza_gcode::gcode_parse;
+//! use cadenza_eval::eval;
+//!
+//! let gcode = "G28\nG1 X100 Y50\n";
+//! let root = gcode_parse(gcode);
+//! // eval() doesn't care that this came from GCode - it's just an AST
+//! ```
+
+use cadenza_syntax::{parse::Parse, token::Kind};
+use rowan::GreenNodeBuilder;
+
+/// Parse GCode source into a Cadenza-compatible AST.
+pub fn gcode_parse(src: &str) -> Parse {
+    GCodeParser::new(src).parse()
+}
+
+struct GCodeParser<'src> {
+    src: &'src str,
+    offset: usize,
+    builder: GreenNodeBuilder<'static>,
+    errors: Vec<cadenza_syntax::parse::ParseError>,
+}
+
+impl<'src> GCodeParser<'src> {
+    fn new(src: &'src str) -> Self {
+        Self {
+            src,
+            offset: 0,
+            builder: GreenNodeBuilder::new(),
+            errors: Vec::new(),
+        }
+    }
+
+    fn parse(mut self) -> Parse {
+        self.builder.start_node(Kind::Root.into());
+
+        for line in self.src.lines() {
+            let line_start = self.offset;
+            self.parse_line(line, line_start);
+            self.offset = line_start + line.len() + 1; // +1 for newline
+        }
+
+        self.builder.finish_node();
+
+        Parse {
+            green: self.builder.finish(),
+            errors: self.errors,
+        }
+    }
+
+    fn parse_line(&mut self, line: &str, line_start: usize) {
+        let line = line.trim();
+
+        if line.is_empty() {
+            return;
+        }
+
+        // Handle comments
+        if line.starts_with(';') {
+            // Comments are trivia, we can skip them or add as whitespace
+            return;
+        }
+
+        // Handle inline comments
+        let code_part = if let Some(pos) = line.find(';') {
+            line[..pos].trim()
+        } else {
+            line
+        };
+
+        if code_part.is_empty() {
+            return;
+        }
+
+        // Parse command as an Apply node
+        self.parse_command(code_part, line_start);
+    }
+
+    fn parse_command(&mut self, code: &str, line_start: usize) {
+        let parts: Vec<&str> = code.split_whitespace().collect();
```

Comment:
We shouldn't be allocating at all. Also this is going to make it difficult to have accurate offsets in the cst using split_whitespace.